docs/modules/publisher-directmethods.md

```diff
@@ -2,34 +2,30 @@
 
 # Configuration via IoT Hub Direct methods
 
-OPC Publisher version 2.8.2 implements the following [IoT Hub Direct Methods](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-direct-methods) which can be called from an application (anywhere in the world) leveraging the [IoT Hub Device SDK](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-sdks).
-
-There are some direct methods which are inherited from 2.5.x in addition to new ones.
-
-Inherited direct methods from 2.5.x:
+OPC Publisher version 2.8.2 implements [IoT Hub Direct Methods](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-direct-methods), which can be called from an application using the [IoT Hub Device SDK](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-sdks).
 
+The following direct methods are exposed:
 - PublishNodes_V1
 - UnpublishNodes_V1
 - UnpublishAllNodes_V1
 - GetConfiguredEndpoints_V1
 - GetConfiguredNodesOnEndpoint_V1
 - GetDiagnosticInfo_V1
-
-New direct methods:
-
 - AddOrUpdateEndpoints_V1
 
-## Terminologies
+If you need to migrate your application from OPC Publisher 2.5.x to OPC Publisher 2.8.2, we provide the needed information in a [separate document](./publisher-migrationpath.md).
+
+## Terminology
 
 The definitions of the important terms used are described below:
 
-- __DataSet__ is a group of NodeIds within an OPC UA Server to be published with the same publishing interval.
-- __DataSetWriter__ has one DataSet and contains the elements required to successfully establish a connection to the OPC UA Server.
-- __DatSetWriterGroup__ is used to group several DataSetWriters for a specific OPC UA server.
+- __DataSet__ is a group of nodes within one OPC UA Server publishing of data value changes of those nodes is done with the same publishing interval.
+- __DataSetWriter__ has one DataSet and contains all information to establish a connection to an OPC UA Server.
+- __DatSetWriterGroup__ is used to group several DataSetWriter's for a specific OPC UA server.
 
 ## Payload Schema
 
-The `_V1` direct methods  uses the  payload schema as described below:
+The `_V1` direct methods use the payload schema as described below:
```

Comment:
Looking at the schema here I am trying to better understand the usage of:

```
"DataSetWriterGroup": "string",
"DataSetWriterId": "string",
"DataSetPublishingInterval": "integer",
```
My current understanding is that there is one DataSetWriterGroup per server Endpoint and I can specify a friendly name for the session (overriding the default hash append behavior).
It looks like I can only specify only ONE DataSetWriterId per Group with ONE DataSetPublishingInterval (as group key?)

Would it be possible to specify the same endpoint multiple times with e.g., the same DataSetWriterGroup but different DataSetWriterId and DataSetPublishingInterval?
e.g.:

```
{
   "EndpointUrl": "opc.tcp://opcplc:50000",
   "DataSetWriterGroup": "Asset0",
   "DataSetWriterId": "DataFlow0",
   "DataSetPublishingInterval": 5000,
   "OpcNodes": [
      {
         "Id": "nsu=http://microsoft.com/Opc/OpcPlc/;s=FastUInt0"
      }
   ],
   "EndpointUrl": "opc.tcp://opcplc:50000",
   "DataSetWriterGroup": "Asset0",
   "DataSetWriterId": "DataFlow1",
   "DataSetPublishingInterval": 1000,
   "OpcNodes": [
      {
         "Id": "nsu=http://microsoft.com/Opc/OpcPlc/;s=FastUInt1"
      }
   ]
}
```

Or even different DataSetWriterGroup for the same endpoint?

If this is possible, what are the use cases/drivers for doing this? 
Are there any implications regarding the throughput or session count one would need to consider?

Reply:
@vslepakov, this is a very good feedback! We need to explain this better.

The schema has certain limitations because we did not want to break the backwards compatibility for the pn.json. 

So basically your example is valid (grouping is a bit broken though) and the result is that the publisher will create 1 session, called `Asset0`. Within this session 2 subscriptions are created for `DataFlow0` and `DataFlow1` having 5000ms respectively 1000ms as publishing intervals.

If 2 DataSetWriterGroup are provided for the same endpoint, like in the example below:
```Json
[
	{
		"EndpointUrl": "opc.tcp://opcplc:50000",
		"DataSetWriterGroup": "Asset0",
		"DataSetWriterId": "DataFlow0",
		"DataSetPublishingInterval": 5000,
		"OpcNodes": [
			{
				"Id": "nsu=http://microsoft.com/Opc/OpcPlc/;s=FastUInt0"
			}
		]
	},
	{
		"EndpointUrl": "opc.tcp://opcplc:50000",
		"DataSetWriterGroup": "Asset1",
		"DataSetWriterId": "DataFlow1",
		"DataSetPublishingInterval": 1000,
		"OpcNodes": [
			{
				"Id": "nsu=http://microsoft.com/Opc/OpcPlc/;s=FastUInt1"
			}
		]
	}
]
```

the result would be that 2 sessions are created, using the very same endpoint & security options, `Asset0` and `Asset1`. Session `Asset0` has one subscription `DataFlow0` publishing 5000ms, respectively session `Asset1` has 1 subscription, `DataFlow1` publishing with 1000ms. 

One of the reasons for introducing these concepts were to give more control to the user on how the sessions/subscriptions/monitored items are created in order to better fit the various server setups we might find in the field. 
Another reason for this change is to introduce PubSub specific aliasing (`DataSetWriterGroup` and `DataSetWriter`) to allow decupling of protocol connectivity settings from logical semantics e.g. if the ip address of an opc ua server changes, but semantically represents the same asset, we can reconfigure the `endpoint url` without reconfiguring the `DataSetWriterGroup` therefore the telemetry format will not change for a seamless vertical consumption. This way redundancy/high availability scenarios can be easier architected. 

Regarding the last question, it is very very difficult to provide generic guidelines when it comes to load balancing sessions/subscriptions/monitored items for an optimal performance between OPC UA Client/server, because of the various profiles OPC UA was designed for. For example, a nano-embedded profile OPC UA Server may only allow 1 session. This is exactly the reason we (optionally) give more control to the user to manage and tune his configuration for best possible outcome. Please note that the new configuration settings are optional, therefore the backwards compatibility is ensured both in terms of configuration as well as behavior.